druid/src/widget/stack.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that arranges its children on top of one another.
+
+use crate::kurbo::Size;
+use crate::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx,
+    UpdateCtx, Widget, WidgetPod,
+};
+
+/// A container that lays out its children along the z-axis, first child at bottom, last child on top.
+#[derive(Default)]
+pub struct Stack<T> {
+    children: Vec<WidgetPod<T, Box<dyn Widget<T>>>>,
+}
+
+impl<T: Data> Stack<T> {
+    /// Create a new stack layout.
+    ///
+    /// The child widgets are laid out on top of one another, from bottom to top.
+    pub fn new() -> Self {
+        Stack {
+            children: Vec::new(),
+        }
+    }
+
+    /// Builder-style variant of `add_child`.
+    ///
+    /// Convenient for assembling a group of widgets in a single expression.
+    pub fn with_child(mut self, child: impl Widget<T> + 'static) -> Self {
+        self.add_child(child);
+        self
+    }
+
+    /// Add a child widget.
+    ///
+    /// See also `with_child`.
+    pub fn add_child(&mut self, child: impl Widget<T> + 'static) {
+        self.children.push(WidgetPod::new(child).boxed());
+    }
+}
+
+impl<T: Data> Widget<T> for Stack<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        match event {
+            Event::MouseDown(e) | Event::MouseUp(e) | Event::MouseMoved(e) => {
+                if let Some(active_child) = self
+                    .children
+                    .iter_mut()
+                    .rev()
+                    .find(|child| child.layout_rect().contains(e.pos))
+                {
```

Comment:
thinking harder this doesn't actually work; it doesn't take account of whether or not a widget is active, and it won't correctly deliver the mouse moved event when the mouse first exits a widget.

I think getting this correct is going to take some fairly tricky logic. One option would be to introduce the idea of a 'hit zone' or a 'hit mask' that lets us represent the obscured portion of a widget. Another option would be to always deliver events to all widgets; and a last option would be to only deliver events to the top widget. 😒

Reply:
One last move move event is sent when mouse is about to leave, right?

Reply:
yes, if the mouse has been over a widget, and moves outside that widget, we deliver the first event that arrives after it has moved.